webkit/LayoutTests/fast/js/select-options-add-expected.txt
This test checks the behavior of the add() method on the select.options object.
It covers both the the one-argument (1.x) and two-argument (2.x) signatures of the add() method.

On success, you will see a series of "PASS" messages, followed by "TEST COMPLETE".


1.1 Add Option to empty Options
PASS select1.options.length is 1
PASS select1.selectedIndex is 0
PASS select1.options[0].value is '1'
PASS select1.options[0].textContent is 'A'

1.2 Add Option to non-empty Options
PASS select1.options.length is 2
PASS select1.selectedIndex is 0
PASS select1.options[0].value is '1'
PASS select1.options[0].textContent is 'A'
PASS select1.options[1].value is '2'
PASS select1.options[1].textContent is 'B'

1.3 Add Option after setting parameters
PASS select1.options.length is 3
PASS select1.selectedIndex is 0
PASS select1.options[0].value is '1'
PASS select1.options[0].textContent is 'A'
PASS select1.options[1].value is '2'
PASS select1.options[1].textContent is 'B'
PASS select1.options[2].value is '3'
PASS select1.options[2].textContent is 'C'

1.4 Add a non-Option element
PASS select1.options.add(option1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select1.options.length is 3
PASS select1.selectedIndex is 0

1.5 Add a non-element (string)
PASS select1.options.add(option1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select1.options.length is 3
PASS select1.selectedIndex is 0

1.6 Add a non-element (number)
PASS select1.options.add(option1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select1.options.length is 3
PASS select1.selectedIndex is 0

1.7 Add a non-element (boolean)
PASS select1.options.add(option1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select1.options.length is 3
PASS select1.selectedIndex is 0

1.8 Add undefined
PASS select1.options.add(option1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select1.options.length is 3
PASS select1.selectedIndex is 0

1.9 Add null
PASS select1.options.add(option1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select1.options.length is 3
PASS select1.selectedIndex is 0

1.10 Add negative infinity
PASS select1.options.add(option1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select1.options.length is 3
PASS select1.selectedIndex is 0

1.11 Add NaN
PASS select1.options.add(option1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select1.options.length is 3
PASS select1.selectedIndex is 0

1.12 Add positive infinity
PASS select1.options.add(option1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select1.options.length is 3
PASS select1.selectedIndex is 0

2.1 Add Option to empty Options
PASS select2.options.length is 1
PASS select2.selectedIndex is 0
PASS select2.options[0].value is '1'
PASS select2.options[0].textContent is 'A'

2.2 Add Option after setting parameters
PASS select2.options.length is 2
PASS select2.selectedIndex is 0
PASS select2.options[0].value is '1'
PASS select2.options[0].textContent is 'A'
PASS select2.options[1].value is '2'
PASS select2.options[1].textContent is 'B'

2.3 Insert Option at beginning of Options
PASS select2.options.length is 3
PASS select2.selectedIndex is 1
PASS select2.options[0].value is '0'
PASS select2.options[0].textContent is 'Z'
PASS select2.options[1].value is '1'
PASS select2.options[1].textContent is 'A'
PASS select2.options[2].value is '2'
PASS select2.options[2].textContent is 'B'

2.4 Insert Option in middle of Options
PASS select2.options.length is 4
PASS select2.selectedIndex is 1
PASS select2.options[0].value is '0'
PASS select2.options[0].textContent is 'Z'
PASS select2.options[1].value is '1'
PASS select2.options[1].textContent is 'A'
PASS select2.options[2].value is '1.5'
PASS select2.options[2].textContent is 'A.5'
PASS select2.options[3].value is '2'
PASS select2.options[3].textContent is 'B'

2.5 Insert Option at end of Options
PASS select2.options.length is 5
PASS select2.selectedIndex is 1
PASS select2.options[0].value is '0'
PASS select2.options[0].textContent is 'Z'
PASS select2.options[1].value is '1'
PASS select2.options[1].textContent is 'A'
PASS select2.options[2].value is '1.5'
PASS select2.options[2].textContent is 'A.5'
PASS select2.options[3].value is '2'
PASS select2.options[3].textContent is 'B'
PASS select2.options[4].value is '3'
PASS select2.options[4].textContent is 'C'

2.6 Insert Option beyond the end of Options
PASS select2.options.length is 6
PASS select2.selectedIndex is 1
PASS select2.options[0].value is '0'
PASS select2.options[0].textContent is 'Z'
PASS select2.options[1].value is '1'
PASS select2.options[1].textContent is 'A'
PASS select2.options[2].value is '1.5'
PASS select2.options[2].textContent is 'A.5'
PASS select2.options[3].value is '2'
PASS select2.options[3].textContent is 'B'
PASS select2.options[4].value is '3'
PASS select2.options[4].textContent is 'C'
PASS select2.options[5].value is '4'
PASS select2.options[5].textContent is 'D'

2.7 Add an Option at index -1
PASS select2.options.length is 7
PASS select2.selectedIndex is 1
PASS select2.options[0].value is '0'
PASS select2.options[0].textContent is 'Z'
PASS select2.options[1].value is '1'
PASS select2.options[1].textContent is 'A'
PASS select2.options[2].value is '1.5'
PASS select2.options[2].textContent is 'A.5'
PASS select2.options[3].value is '2'
PASS select2.options[3].textContent is 'B'
PASS select2.options[4].value is '3'
PASS select2.options[4].textContent is 'C'
PASS select2.options[5].value is '4'
PASS select2.options[5].textContent is 'D'
PASS select2.options[6].value is '5'
PASS select2.options[6].textContent is 'E'

2.8 Add an Option at index -2
PASS select2.options.add(option2, -2) threw exception Error: INDEX_SIZE_ERR: DOM Exception 1.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.9 Add an Option at index -Infinity
PASS select2.options.add(option2, -1/0) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.10 Add an Option at index NaN
PASS select2.options.add(option2, 0/0) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.11 Add an Option at index Infinity
PASS select2.options.add(option2, 1/0) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.12 Add a non-Option element
PASS select2.options.add(option2, 1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.13 Add a non-element (string)
PASS select2.options.add(option2, 1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.14 Add a non-element (number)
PASS select2.options.add(option2, 1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.15 Add a non-element (boolean)
PASS select2.options.add(option2, 1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.16 Add undefined
PASS select2.options.add(option2, 1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.17 Add null
PASS select2.options.add(option2, 1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.18 Add negative infinity
PASS select2.options.add(option2, 1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.19 Add NaN
PASS select2.options.add(option2, 1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

2.20 Add positive infinity
PASS select2.options.add(option2, 1) threw exception Error: TYPE_MISMATCH_ERR: DOM Exception 17.
PASS select2.options.length is 7
PASS select2.selectedIndex is 1

PASS successfullyParsed is true

TEST COMPLETE



webkit/LayoutTests/fast/js/select-options-remove-expected.txt
This test checks the behavior of the remove() method on the select.options object.

On success, you will see a series of "PASS" messages, followed by "TEST COMPLETE".


1.1 Remove (object) from empty Options
PASS select1.options.remove(value) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.2 Remove (string) from empty Options
PASS select1.options.remove(value) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.3 Remove (float) from empty Options
PASS select1.options.remove(value) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.4 Remove (boolean) from empty Options
PASS select1.options.remove(value) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.5 Remove (undefined) from empty Options
PASS select1.options.remove(value) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.6 Remove (null) from empty Options
PASS select1.options.remove(value) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.7 Remove (negative infinity) from empty Options
PASS select1.options.remove(value) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.8 Remove (NaN) from empty Options
PASS select1.options.remove(value) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.9 Remove (positive infinity) from empty Options
PASS select1.options.remove(value) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.10 Remove no args from empty Options
PASS select1.options.remove() is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.11 Remove too many args from empty Options
PASS select1.options.remove(0, 'foo') is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.12 Remove invalid index -2 from empty Options
PASS select1.options.remove(-2) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.13 Remove invalid index -1 from empty Options
PASS select1.options.remove(-1) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.14 Remove index 0 from empty Options
PASS select1.options.remove(0) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

1.15 Remove index 1 from empty Options
PASS select1.options.remove(1) is undefined
PASS select1.options.length is 0
PASS select1.selectedIndex is -1

2.1 Remove (object) from non-empty Options
PASS select2.options.remove(value) is undefined
PASS select2.options.length is 15
PASS select2.selectedIndex is 13
PASS select2.options[0].value is 'B'

2.2 Remove (string) from non-empty Options
PASS select2.options.remove(value) is undefined
PASS select2.options.length is 14
PASS select2.selectedIndex is 12
PASS select2.options[0].value is 'C'

2.3 Remove (float) from non-empty Options
PASS select2.options.remove(value) is undefined
PASS select2.options.length is 13
PASS select2.selectedIndex is 11
PASS select2.options[3].value is 'G'

2.4 Remove (boolean true) from non-empty Options
PASS select2.options.remove(value) is undefined
PASS select2.options.length is 12
PASS select2.selectedIndex is 10
PASS select2.options[1].value is 'E'

2.5 Remove (boolean false) from non-empty Options
PASS select2.options.remove(value) is undefined
PASS select2.options.length is 11
PASS select2.selectedIndex is 9
PASS select2.options[1].value is 'G'

2.6 Remove (undefined) from non-empty Options
PASS select2.options.remove(value) is undefined
PASS select2.options.length is 10
PASS select2.selectedIndex is 8
PASS select2.options[0].value is 'G'

2.7 Remove (null) from non-empty Options
PASS select2.options.remove(value) is undefined
PASS select2.options.length is 9
PASS select2.selectedIndex is 7
PASS select2.options[0].value is 'H'

2.8 Remove (negative infinity) from non-empty Options
PASS select2.options.remove(value) is undefined
PASS select2.options.length is 8
PASS select2.selectedIndex is 6
PASS select2.options[0].value is 'I'

2.9 Remove (NaN) from non-empty Options
PASS select2.options.remove(value) is undefined
PASS select2.options.length is 7
PASS select2.selectedIndex is 5
PASS select2.options[0].value is 'J'

2.10 Remove (positive infinity) from non-empty Options
PASS select2.options.remove(value) is undefined
PASS select2.options.length is 6
PASS select2.selectedIndex is 4
PASS select2.options[0].value is 'K'

2.11 Remove no args from non-empty Options
PASS select2.options.remove() is undefined
PASS select2.options.length is 5
PASS select2.selectedIndex is 3
PASS select2.options[0].value is 'L'

2.12 Remove too many args from non-empty Options
PASS select2.options.remove(0, 'foo') is undefined
PASS select2.options.length is 4
PASS select2.selectedIndex is 2
PASS select2.options[0].value is 'M'

2.13 Remove invalid index -2 from non-empty Options
PASS select2.options.remove(-2) is undefined
PASS select2.options.length is 4
PASS select2.selectedIndex is 2
PASS select2.options[2].value is 'O'

2.14 Remove invalid index -1 from non-empty Options
PASS select2.options.remove(-1) is undefined
PASS select2.options.length is 4
PASS select2.selectedIndex is 2
PASS select2.options[3].value is 'P'

2.15 Remove index 0 from non-empty Options
PASS select2.options.remove(0) is undefined
PASS select2.options.length is 3
PASS select2.selectedIndex is 1
PASS select2.options[0].value is 'N'

2.16 Remove index 1 from non-empty Options
PASS select2.options.remove(1) is undefined
PASS select2.options.length is 2
PASS select2.selectedIndex is 0
PASS select2.options[1].value is 'P'

PASS successfullyParsed is true

TEST COMPLETE
 


webkit/ChangeLog
2010-09-14  Andras Becsi  <abecsi@webkit.org>

        Reviewed by Csaba Osztrogonác.

        [Qt] qmake doesn't generate correct dependencies for rcc resource files
        https://bugs.webkit.org/show_bug.cgi?id=45747

        * WebKit.pri: add OUT_PWD to DEPENDPATH

2010-09-11  Simon Hausmann  <simon.hausmann@nokia.com>

        Reviewed by Andreas Kling.

        [Qt] V8 port: webkit project files changes
        https://bugs.webkit.org/show_bug.cgi?id=45140

        * WebKit.pro: Don't compile JavaScriptCore and jsc when
        building with v8.

2010-09-11  Xan Lopez  <xlopez@igalia.com>

        Reviewed by Martin Robinson.

        [GTK] Make introspection work with g-o-i 0.9.5
        https://bugs.webkit.org/show_bug.cgi?id=45590

        Bump required gobject-introspection version to 0.9.5.

        * configure.ac:

2010-09-09  Benjamin Poulain  <benjamin.poulain@nokia.com>

        Unreviewed buildfix.

        [Qt] Treat warnings as errors by default for gcc
        https://bugs.webkit.org/show_bug.cgi?id=43191

        The buildfix r66990 fixes the build for the Maemo
        branch of Qt, but the symbols maemo5 and 6 are not defined
        for regular branches on ARM.
        This fixes the build by disabling -Werror for any ARM platform.

        * WebKit.pri: Don't add -Werror for arm platforms.

2010-09-08  Peter Kasting  <pkasting@google.com>

        Reviewed by David Hyatt.

        Add smooth scrolling framework, and a Windows implementation.
        https://bugs.webkit.org/show_bug.cgi?id=32356

        * wscript: Add Windows ScrollAnimator.

2010-09-08  Csaba Osztrogonác  <ossy@webkit.org>

        Unreviewed buildfix after r66972.

        [Qt] Treat warnings as errors by default for gcc
        https://bugs.webkit.org/show_bug.cgi?id=43191

        * WebKit.pri: Don't add -Werror for maemo platforms.

2010-09-08  Martin Robinson  <mrobinson@igalia.com>

        Reviewed by Xan Lopez.

        [GTK] Need a WebSocket implementation
        https://bugs.webkit.org/show_bug.cgi?id=45197

        * configure.ac: Enable WebSocket by default.

2010-09-08  Gyuyoung Kim  <gyuyoung.kim@samsung.com>

        Reviewed by Kenneth Rohde Christiansen.

        [EFL] Support to enable HTML5's Video based on gstreamer in WebKit-EFL
        https://bugs.webkit.org/show_bug.cgi?id=44098

        To support video of HTML5 based on gstreamer, add cmake files for gstreamer. 

        * cmake/FindGStreamer-App.cmake: Added.
        * cmake/FindGStreamer-Base.cmake: Added.
        * cmake/FindGStreamer-Interfaces.cmake: Added.
        * cmake/FindGStreamer-Pbutils.cmake: Added.
        * cmake/FindGStreamer-Plugins-Base.cmake: Added.
        * cmake/FindGStreamer-Video.cmake: Added.
        * cmake/FindGStreamer.cmake: Added.
        * cmake/OptionsEfl.cmake:

2010-09-08  Csaba Osztrogonác  <ossy@webkit.org>

        Reviewed by Andreas Kling.

        [Qt] Treat warnings as errors by default for gcc
        https://bugs.webkit.org/show_bug.cgi?id=43191

        * WebKit.pri: -Werror added to QMAKE_CXXFLAGS for x86 Linux platforms

2010-09-07  Arno Renevier  <arno@renevier.net>

        Reviewed by Holger Freyther.

        gitignore *~ files and gtk *.pot files
        https://bugs.webkit.org/show_bug.cgi?id=45129

        * .gitignore:

2010-09-04  Lucas De Marchi  <lucas.demarchi@profusion.mobi>

        Reviewed by Kenneth Rohde Christiansen.

        [EFL] Move test browser to WebKitTools directory
        https://bugs.webkit.org/show_bug.cgi?id=45212

        Follow other ports like QT and GTK which moved the test browser to
        WebKitTools directory.

        * CMakeLists.txt: Include test browser if port defines one.

2010-09-02  Lucas De Marchi  <lucas.demarchi@profusion.mobi>

        Reviewed by Antonio Gomes.

        [EFL] Name install directories according to library name
        https://bugs.webkit.org/show_bug.cgi?id=45126

        * cmake/OptionsEfl.cmake: Use library name in order to define the data
        directory.

2010-09-02  Lucas De Marchi  <lucas.demarchi@profusion.mobi>

        Reviewed by Antonio Gomes.

        [EFL] Do not override custom compile flags
        https://bugs.webkit.org/show_bug.cgi?id=45125

        Set a default build type if and only if user did not define one as
        command line options and he did not give custom CFLAGS or CXXFLAGS.
        Otherwise, flags from default build type would override user-defined
        ones.

        * CMakeLists.txt:

2010-09-02  Lucas De Marchi  <lucas.demarchi@profusion.mobi>

        Reviewed by Antonio Gomes.

        [EFL] Fix double addition of -fPIC
        https://bugs.webkit.org/show_bug.cgi?id=45122

         When SHARED_CORE=ON, -fPIC is automatically added because all
         libraries are dynamic. Only when static libraries are built that
         -fPIC needs to be manually inserted (because the final library,
         libewebkit.so, is dynamic).

        * cmake/OptionsCommon.cmake: Remove -fPIC flag.
        * cmake/WebKitHelpers.cmake: Add -fPIC flags iff SHARED_CORE=ON.

2010-09-01  Ryuan Choi  <ryuan.choi@samsung.com>

        Reviewed by Antonio Gomes.

        [EFL] Need to check LibSoup version
        https://bugs.webkit.org/show_bug.cgi?id=44658

        Add version check of LibSoup.

        * cmake/FindLibSoup2.cmake:

2010-08-31  Gustavo Noronha Silva  <gustavo.noronha@collabora.co.uk>

        Reviewed by Martin Robinson.

        Preparations for the 1.3.4 release. Bump webkit version to 543.7,
        as well.

        * configure.ac:

2010-08-30  Lucas De Marchi  <lucas.demarchi@profusion.mobi>

        Reviewed by Adam Barth.

        [EFL] Remove compiler optimization for gcc 4.5.1
        https://bugs.webkit.org/show_bug.cgi?id=44520


webkit/BugsSite/Bugzilla/Config.pm
# -*- Mode: perl; indent-tabs-mode: nil -*-
#
# The contents of this file are subject to the Mozilla Public
# License Version 1.1 (the "License"); you may not use this file
# except in compliance with the License. You may obtain a copy of
# the License at http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS
# IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
# implied. See the License for the specific language governing
# rights and limitations under the License.
#
# The Original Code is the Bugzilla Bug Tracking System.
#
# The Initial Developer of the Original Code is Netscape Communications
# Corporation. Portions created by Netscape are
# Copyright (C) 1998 Netscape Communications Corporation. All
# Rights Reserved.
#
# Contributor(s): Terry Weissman <terry@mozilla.org>
#                 Dawn Endico <endico@mozilla.org>
#                 Dan Mosedale <dmose@mozilla.org>
#                 Joe Robins <jmrobins@tgix.com>
#                 Jake <jake@bugzilla.org>
#                 J. Paul Reed <preed@sigkill.com>
#                 Bradley Baetz <bbaetz@student.usyd.edu.au>
#                 Christopher Aillon <christopher@aillon.com>
#                 Erik Stambaugh <erik@dasbistro.com>
#                 Frédéric Buclin <LpSolit@gmail.com>

package Bugzilla::Config;

use strict;

use base qw(Exporter);
use Bugzilla::Constants;
use Data::Dumper;
use File::Temp;

# Don't export localvars by default - people should have to explicitly
# ask for it, as a (probably futile) attempt to stop code using it
# when it shouldn't
%Bugzilla::Config::EXPORT_TAGS =
  (
   admin => [qw(update_params SetParam write_params)],
  );
Exporter::export_ok_tags('admin');

use vars qw(@param_list);

# INITIALISATION CODE
# Perl throws a warning if we use bz_locations() directly after do.
our %params;
# Load in the param definitions
sub _load_params {
    my $panels = param_panels();
    foreach my $panel (keys %$panels) {
        my $module = $panels->{$panel};
        eval("require $module") || die $@;
        my @new_param_list = "$module"->get_param_list();
        foreach my $item (@new_param_list) {
            $params{$item->{'name'}} = $item;
        }
        push(@param_list, @new_param_list);
    }
}
# END INIT CODE

# Subroutines go here

sub param_panels {
    my $param_panels = {};
    my $libpath = bz_locations()->{'libpath'};
    foreach my $item ((glob "$libpath/Bugzilla/Config/*.pm")) {
        $item =~ m#/([^/]+)\.pm$#;
        my $module = $1;
        $param_panels->{$module} = "Bugzilla::Config::$module" unless $module eq 'Common';
    }
    # Now check for any hooked params
    Bugzilla::Hook::process('config', { config => $param_panels });
    return $param_panels;
}

sub SetParam {
    my ($name, $value) = @_;

    _load_params unless %params;
    die "Unknown param $name" unless (exists $params{$name});

    my $entry = $params{$name};

    # sanity check the value

    # XXX - This runs the checks. Which would be good, except that
    # check_shadowdb creates the database as a side effect, and so the
    # checker fails the second time around...
    if ($name ne 'shadowdb' && exists $entry->{'checker'}) {
        my $err = $entry->{'checker'}->($value, $entry);
        die "Param $name is not valid: $err" unless $err eq '';
    }

    Bugzilla->params->{$name} = $value;
}

sub update_params {
    my ($params) = @_;
    my $answer = Bugzilla->installation_answers;

    my $param = read_param_file();

    # If we didn't return any param values, then this is a new installation.
    my $new_install = !(keys %$param);

    # --- UPDATE OLD PARAMS ---

    # Old Bugzilla versions stored the version number in the params file
    # We don't want it, so get rid of it
    delete $param->{'version'};

    # Change from usebrowserinfo to defaultplatform/defaultopsys combo
    if (exists $param->{'usebrowserinfo'}) {
        if (!$param->{'usebrowserinfo'}) {
            if (!exists $param->{'defaultplatform'}) {
                $param->{'defaultplatform'} = 'Other';
            }
            if (!exists $param->{'defaultopsys'}) {
                $param->{'defaultopsys'} = 'Other';
            }
        }
        delete $param->{'usebrowserinfo'};
    }

    # Change from a boolean for quips to multi-state
    if (exists $param->{'usequip'} && !exists $param->{'enablequips'}) {
        $param->{'enablequips'} = $param->{'usequip'} ? 'on' : 'off';
        delete $param->{'usequip'};
    }

    # Change from old product groups to controls for group_control_map
    # 2002-10-14 bug 147275 bugreport@peshkin.net
    if (exists $param->{'usebuggroups'} && 
        !exists $param->{'makeproductgroups'}) 
    {
        $param->{'makeproductgroups'} = $param->{'usebuggroups'};
    }
    if (exists $param->{'usebuggroupsentry'} 
       && !exists $param->{'useentrygroupdefault'}) {
        $param->{'useentrygroupdefault'} = $param->{'usebuggroupsentry'};
    }

    # Modularise auth code
    if (exists $param->{'useLDAP'} && !exists $param->{'loginmethod'}) {
        $param->{'loginmethod'} = $param->{'useLDAP'} ? "LDAP" : "DB";
    }

    # set verify method to whatever loginmethod was
    if (exists $param->{'loginmethod'} 
        && !exists $param->{'user_verify_class'}) 
    {
        $param->{'user_verify_class'} = $param->{'loginmethod'};
        delete $param->{'loginmethod'};
    }

    # Remove quip-display control from parameters
    # and give it to users via User Settings (Bug 41972)
    if ( exists $param->{'enablequips'} 
         && !exists $param->{'quip_list_entry_control'}) 
    {
        my $new_value;
        ($param->{'enablequips'} eq 'on')       && do {$new_value = 'open';};
        ($param->{'enablequips'} eq 'approved') && do {$new_value = 'moderated';};
        ($param->{'enablequips'} eq 'frozen')   && do {$new_value = 'closed';};
        ($param->{'enablequips'} eq 'off')      && do {$new_value = 'closed';};
        $param->{'quip_list_entry_control'} = $new_value;
        delete $param->{'enablequips'};
    }

    # Old mail_delivery_method choices contained no uppercase characters
    if (exists $param->{'mail_delivery_method'}
        && $param->{'mail_delivery_method'} !~ /[A-Z]/) {
        my $method = $param->{'mail_delivery_method'};
        my %translation = (
            'sendmail' => 'Sendmail',
            'smtp'     => 'SMTP',
            'qmail'    => 'Qmail',
            'testfile' => 'Test',
            'none'     => 'None');
        $param->{'mail_delivery_method'} = $translation{$method};
    }

    # --- DEFAULTS FOR NEW PARAMS ---

    _load_params unless %params;
    foreach my $item (@param_list) {
        my $name = $item->{'name'};
        unless (exists $param->{$name}) {
            print "New parameter: $name\n" unless $new_install;
            $param->{$name} = $answer->{$name} || $item->{'default'};
        }
    }

    $param->{'utf8'} = 1 if $new_install;

    # --- REMOVE OLD PARAMS ---

    my %oldparams;
    # Remove any old params, put them in old-params.txt
    foreach my $item (keys %$param) {
        if (!grep($_ eq $item, map ($_->{'name'}, @param_list))) {
            $oldparams{$item} = $param->{$item};
            delete $param->{$item};
        }
    }

    if (scalar(keys %oldparams)) {
        my $op_file = new IO::File('old-params.txt', '>>', 0600)
          || die "old-params.txt: $!";

        print "The following parameters are no longer used in Bugzilla,",
              " and so have been\nmoved from your parameters file into",
              " old-params.txt:\n";

        local $Data::Dumper::Terse  = 1;
        local $Data::Dumper::Indent = 0;

        my $comma = "";
        foreach my $item (keys %oldparams) {
            print $op_file "\n\n$item:\n" . Data::Dumper->Dump([$oldparams{$item}]) . "\n";
            print "${comma}$item";
            $comma = ", ";
        }
        print "\n";
        $op_file->close;
    }

    if (ON_WINDOWS && !-e SENDMAIL_EXE
        && $param->{'mail_delivery_method'} eq 'Sendmail')
    {
        my $smtp = $answer->{'SMTP_SERVER'};
        if (!$smtp) {
            print "\nBugzilla requires an SMTP server to function on",
                  " Windows.\nPlease enter your SMTP server's hostname: ";
            $smtp = <STDIN>;
            chomp $smtp;
            if ($smtp) {
                $param->{'smtpserver'} = $smtp;
             }
             else {
                print "\nWarning: No SMTP Server provided, defaulting to",
                      " localhost\n";
            }
        }

        $param->{'mail_delivery_method'} = 'SMTP';
    }

    write_params($param);

    # Return deleted params and values so that checksetup.pl has a chance
    # to convert old params to new data.
    return %oldparams;
}

sub write_params {
    my ($param_data) = @_;
    $param_data ||= Bugzilla->params;

    my $datadir    = bz_locations()->{'datadir'};
    my $param_file = "$datadir/params";

    local $Data::Dumper::Sortkeys = 1;

    my ($fh, $tmpname) = File::Temp::tempfile('params.XXXXX',
                                              DIR => $datadir );

    print $fh (Data::Dumper->Dump([$param_data], ['*param']))
      || die "Can't write param file: $!";

    close $fh;

    rename $tmpname, $param_file
      or die "Can't rename $tmpname to $param_file: $!";

    ChmodDataFile($param_file, 0666);

    # And now we have to reset the params cache so that Bugzilla will re-read
    # them.
    delete Bugzilla->request_cache->{params};
}

# Some files in the data directory must be world readable if and only if
# we don't have a webserver group. Call this function to do this.
# This will become a private function once all the datafile handling stuff
# moves into this package

# This sub is not perldoc'd for that reason - noone should know about it
sub ChmodDataFile {
    my ($file, $mask) = @_;
    my $perm = 0770;
    if ((stat(bz_locations()->{'datadir'}))[2] & 0002) {
        $perm = 0777;
    }
    $perm = $perm & $mask;
    chmod $perm,$file;
}

sub read_param_file {
    my %params;
    my $datadir = bz_locations()->{'datadir'};
    if (-e "$datadir/params") {
        # Note that checksetup.pl sets file permissions on '$datadir/params'

        # Using Safe mode is _not_ a guarantee of safety if someone does
        # manage to write to the file. However, it won't hurt...
        # See bug 165144 for not needing to eval this at all
        my $s = new Safe;

        $s->rdo("$datadir/params");
        die "Error reading $datadir/params: $!" if $!;
        die "Error evaluating $datadir/params: $@" if $@;

        # Now read the param back out from the sandbox
        %params = %{$s->varglob('param')};
    }
    elsif ($ENV{'SERVER_SOFTWARE'}) {
       # We're in a CGI, but the params file doesn't exist. We can't
       # Template Toolkit, or even install_string, since checksetup
       # might not have thrown an error. Bugzilla::CGI->new
       # hasn't even been called yet, so we manually use CGI::Carp here
       # so that the user sees the error.
       require CGI::Carp;
       CGI::Carp->import('fatalsToBrowser');
       die "The $datadir/params file does not exist."
           . ' You probably need to run checksetup.pl.',
    }
    return \%params;
}

1;

__END__

=head1 NAME

Bugzilla::Config - Configuration parameters for Bugzilla

=head1 SYNOPSIS

  # Administration functions
  use Bugzilla::Config qw(:admin);

  update_params();
  SetParam($param, $value);
  write_params();

=head1 DESCRIPTION

This package contains ways to access Bugzilla configuration parameters.

=head1 FUNCTIONS

=head2 Parameters

Parameters can be set, retrieved, and updated.

=over 4

=item C<SetParam($name, $value)>

Sets the param named $name to $value. Values are checked using the checker
function for the given param if one exists.

=item C<update_params()>

Updates the parameters, by transitioning old params to new formats, setting
defaults for new params, and removing obsolete ones. Used by F<checksetup.pl>
in the process of an installation or upgrade.

Prints out information about what it's doing, if it makes any changes.

May prompt the user for input, if certain required parameters are not
specified.

=item C<write_params($params)>

Description: Writes the parameters to disk.

Params:      C<$params> (optional) - A hashref to write to the disk
               instead of C<Bugzilla->params>. Used only by
               C<update_params>.

Returns:     nothing

=item C<read_param_file()>

Description: Most callers should never need this. This is used
             by C<Bugzilla->params> to directly read C<$datadir/params>
             and load it into memory. Use C<Bugzilla->params> instead.

Params:      none

Returns:     A hashref containing the current params in C<$datadir/params>.

=back
